pruning/manager.go

```diff
@@ -0,0 +1,216 @@
+				next = e.Next()
+				m.pruneSnapshotHeights.Remove(e)
+			} else {
+				next = e.Next()
+			}
+		}
+	}()
+
+	if int64(m.opts.KeepRecent) < previousHeight {
+		pruneHeight := previousHeight - int64(m.opts.KeepRecent)
+		// We consider this height to be pruned iff:
+		//
+		// - snapshotInterval is zero as that means that all heights should be pruned.
+		// - snapshotInterval % (height - KeepRecent) != 0 as that means the height is not
+		// a 'snapshot' height.
+		if m.snapshotInterval == 0 || pruneHeight%int64(m.snapshotInterval) != 0 {
+			m.pruneHeights = append(m.pruneHeights, pruneHeight)
+			return pruneHeight
+		}
+	}
+	return 0
+}
+
+func (m *Manager) HandleHeightSnapshot(height int64) {
+	if m.opts.GetPruningStrategy() == types.PruningNothing {
+		return
+	}
+	m.mx.Lock()
+	defer m.mx.Unlock()
+	m.logger.Debug("HandleHeightSnapshot", "height", height)
+	m.pruneSnapshotHeights.PushBack(height)
+}
+
+// SetSnapshotInterval sets the interval at which the snapshots are taken.
+func (m *Manager) SetSnapshotInterval(snapshotInterval uint64) {
+	m.snapshotInterval = snapshotInterval
+}
+
+// ShouldPruneAtHeight return true if the given height should be pruned, false otherwise
+func (m *Manager) ShouldPruneAtHeight(height int64) bool {
+	return m.opts.GetPruningStrategy() != types.PruningNothing && m.opts.Interval > 0 && height%int64(m.opts.Interval) == 0
+}
+
+// FlushPruningHeights flushes the pruning heights to the database for crash recovery.
+func (m *Manager) FlushPruningHeights(batch dbm.Batch) {
+	if m.opts.GetPruningStrategy() == types.PruningNothing {
+		return
+	}
+	m.flushPruningHeights(batch)
+	m.flushPruningSnapshotHeights(batch)
+}
+
+// LoadPruningHeights loads the pruning heights from the database as a crash recovery.
+func (m *Manager) LoadPruningHeights(db dbm.DB) error {
+	if m.opts.GetPruningStrategy() == types.PruningNothing {
+		return nil
+	}
+	if err := m.loadPruningHeights(db); err != nil {
+		return err
+	}
+	if err := m.loadPruningSnapshotHeights(db); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *Manager) loadPruningHeights(db dbm.DB) error {
+	bz, err := db.Get([]byte(pruneHeightsKey))
+	if err != nil {
+		return fmt.Errorf("failed to get pruned heights: %w", err)
+	}
+	if len(bz) == 0 {
+		return nil
+	}
+
+	prunedHeights := make([]int64, len(bz)/8)
+	i, offset := 0, 0
+	for offset < len(bz) {
+		prunedHeights[i] = int64(binary.BigEndian.Uint64(bz[offset : offset+8]))
+		i++
+		offset += 8
+	}
+
+	if len(prunedHeights) > 0 {
+		m.pruneHeights = prunedHeights
+	}
+
+	return nil
+}
+
+func (m *Manager) loadPruningSnapshotHeights(db dbm.DB) error {
+	bz, err := db.Get([]byte(pruneSnapshotHeightsKey))
+	if err != nil {
+		return fmt.Errorf("failed to get post-snapshot pruned heights: %w", err)
+	}
+	if len(bz) == 0 {
+		return nil
+	}
+
+	pruneSnapshotHeights := list.New()
+	i, offset := 0, 0
+	for offset < len(bz) {
+		pruneSnapshotHeights.PushBack(int64(binary.BigEndian.Uint64(bz[offset : offset+8])))
```

Comment:
There is potential to incur under flow here and save negative heights if someone tampered with the database, which is an attack surface I'd consider. I say this because when saving these heights to the database we save them as Uint64 big endian serialized values, and don't check if negative heights were saved. However to defensively operate, in RAM we should be validating that the values we retrieved are not under flows. I'd encourage doing this:

```go
h := int64(binary.BigEndian.Uint64(bz[offset : offset + 8]))
if h < 0 {
    return fmt.Errorf("loaded negative height: %d", h)
}
```